usim/py/events.py

```diff
@@ -0,0 +1,642 @@
+    # pair will evaluate concurrently, and is *not* defused on error.
+    # So we *must* preserve all intermediate events to defuse them in one go.
+    def __or__(self, other):
+        if isinstance(other, Event):
+            return AnyOf(self.env, (self, other))
+        return NotImplemented
+
+    def __and__(self, other):
+        if isinstance(other, Event):
+            return AllOf(self.env, (self, other))
+        return NotImplemented
+
+    def __await__(self):
+        yield from self.__usimpy_flag__.__await__()
+        result, error = self._value
+        if error is not None:
+            # the waiter will handle our exception
+            self.defused = True
+            raise error
+        else:
+            return result  # noqa: B901
+
+    async def _invoke_callbacks(self):
+        # simpy does this in core.Environment.step
+        # we neither have step, nor a callback event loop
+        assert self._value is not None,\
+            f'{self.__class__.__name__} must be triggered before invoking callbacks'
+        assert self.callbacks is not None,\
+            f'{self.__class__.__name__} can invoke callbacks only once'
+        callbacks, self.callbacks = self.callbacks, None
+        for callback in callbacks:  # type: Event
+            callback.trigger(self)
+        value, exception = self._value
+        if exception is not None and not self.defused:
+            raise exception
+
+    async def __usimpy_schedule__(self):
+        """Coroutine to schedule this Event in ``usim.py``"""
+        await self._invoke_callbacks()
+
+    def _trigger(self):
+        """Awake all waiting tasks and schedule the event itself"""
+        self.__usimpy_flag__._value = True
+        self.__usimpy_flag__.__trigger__()
+        self.env.schedule(self)
+
+    @property
+    def triggered(self) -> bool:
+        """Whether this event is being or has been processed"""
+        return bool(self.__usimpy_flag__)
+
+    @property
+    def processed(self) -> bool:
+        """Whether the callbacks have been processed"""
+        return self.callbacks is None
+
+    @property
+    def ok(self) -> bool:
+        """
+        Whether this event has been processed successfully
+
+        .. hint::
+
+            Migrate by using ``bool(flag)`` instead.
+        """
+        return self._value is not None and self._value[1] is None
+
+    @property
+    def value(self) -> Union[V, Exception]:
+        """The value of the event if it has been triggered"""
+        try:
+            value, exception = self._value
+        except TypeError:
+            raise AttributeError(
+                f"{self.__class__.__name__!r} object has no attribute 'value' set yet"
+            )
+        else:
+            return value if exception is None else exception
+
+    def trigger(self, event: 'Event') -> 'Event':
+        """
+        Trigger this event to match the state of ``event``
```

Comment:
What do you mean with *match the state of event*?

Reply:
In SimPy, ``self`` copies the result and several flags from ``event``. In ``usim.py``, this is all folded into the result. How about this?

```suggestion
        Trigger this event and inherit the value or exception of ``event``
```